exchangeBridge/exchange.go

```diff
@@ -0,0 +1,161 @@
+package exchangeBridge
+
+import (
+	"context"
+
+	// We need the typeurl register calls that occur in the init of this package in order to be able to unmarshal events
+	// in the RepublishEvents func below
+	_ "github.com/containerd/containerd/api/events"
+
+	// Even though we are following the v2 runtime model, we are currently re-using a struct definition (Envelope) from
+	// the v1 event API
+	eventapi "github.com/containerd/containerd/api/services/events/v1"
+
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/ttrpc"
+	"github.com/containerd/typeurl"
+	"github.com/gogo/protobuf/types"
+)
+
+const (
+	ExchangeBridgeServiceName = "aws.firecracker.containerd.exchangeBridge"
+	GetEventMethodName = "GetEvent"
+)
+
+// The ExchangeBridge interface provides an API for retrieving containerd events from a remote event exchange. It exists
+// separately from containerd's existing EventClient and EventServer interfaces in order to support retrieving events
+// via long-polling as opposed to a streaming model. This allows us to use TTRPC for the implementation, which currently
+// does not support streaming.
+type ExchangeBridge interface {
+	GetEvent(ctx context.Context) (*eventapi.Envelope, error)
+}
+
+type exchangeBridgeService struct {
+	eventChan <-chan *events.Envelope
+	errChan   <-chan error
+}
+
+// NewExchangeBridgeService returns a server-side implementation of the ExchangeBridge interface. Given an existing
+// exchange, it will subscribe to that exchange with no filters. Each GetEvent call pops and returns an event buffered
+// in the subscription's channel of published events.
+func NewExchangeBridgeService(ctx context.Context, eventSource events.Subscriber) ExchangeBridge {
+	eventChan, errChan := eventSource.Subscribe(ctx)
+	return &exchangeBridgeService{
+		eventChan: eventChan,
+		errChan:   errChan,
+	}
+}
+
+func RegisterExchangeBridgeService(srv *ttrpc.Server, svc ExchangeBridge) {
+	srv.Register(ExchangeBridgeServiceName, map[string]ttrpc.Method{
+		GetEventMethodName: func(ctx context.Context, unmarshal func(interface{}) error) (interface{}, error) {
+			return svc.GetEvent(ctx)
+		},
+	})
+}
+
+func (s *exchangeBridgeService) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var (
+		envelope *eventapi.Envelope
+		err error
+	)
+
+	select {
+	case receivedEnvelope := <-s.eventChan:
+		envelope = &eventapi.Envelope{
+			Timestamp: receivedEnvelope.Timestamp,
+			Namespace: receivedEnvelope.Namespace,
+			Topic:     receivedEnvelope.Topic,
+			Event:     receivedEnvelope.Event,
+		}
+	case err = <-s.errChan:
```

Comment:
should we return here?

Reply:
We don't need to as we will return the err below. The implementation here works as is but I agree it's kind of a different approach than what I did elsewhere, so I can update this to just return in each select case for the sake of consistency.

Reply:
I prefer to return as early as possible. This makes reading code a bit easier.

Reply:
I'd encourage returning at the site of an error.  It's easy to miss that the error is supposed to be returned when we come back and edit this code later if it's only returned at the bottom of the function.  I'd also encourage returning `nil, err` so that we explicitly have no `eventapi.Envelope` instead of one that may be invalid.

Reply:
Yeah that makes total sense, will update